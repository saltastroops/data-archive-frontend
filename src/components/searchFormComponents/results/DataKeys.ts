/**
 * Constants for data keys used in the search results table.
 */
const DataKeys = {
  CART: "cart",
  DECLINATION: "declination",
  FILENAME: "filename",
  INSTRUMENT: "instrument",
  OBSERVATION_NAME: "observationName",
  PREVIEW_IMAGE_URL: "url",
  PROPOSAL_CODE: "proposalCode",
  RIGHT_ASCENSION: "rightAscension",
  TARGET_NAME: "targetName",
  TELESCOPE: "telescope"
};

export default DataKeys;
