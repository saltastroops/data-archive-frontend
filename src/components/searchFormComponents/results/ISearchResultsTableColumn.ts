/**
 * A search results table column.
 *
 * Properties:
 * -----------
 * dataKey:
 *     A string identifying the type of data in this column.
 * name:
 *     A human-friendly column name.
 * visible:
 *     Whether the column should be visible.
 */
export default interface ISearchResultsTableColumn {
  dataKey: string;
  name: string;
  visible: boolean;
}
