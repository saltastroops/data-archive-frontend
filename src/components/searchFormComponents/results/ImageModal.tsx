import * as React from "react";
import Modal from "react-responsive-modal";

interface IImageModal {
  image: {
    url: string;
    alt: string;
  };
  open: boolean;
  closeModal: () => void;
}

export default (props: IImageModal) => {
  const { open, closeModal } = props;
  const { url, alt } = props.image;

  return (
    <Modal open={open} onClose={closeModal} center={true}>
      <figure className="image is-2by2">
        <img className="card-image" id="img01" src={url} alt={alt} />{" "}
        <p className={"notification"}>{alt}</p>
      </figure>
    </Modal>
  );
};
