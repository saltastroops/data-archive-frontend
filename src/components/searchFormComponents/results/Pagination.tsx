import * as React from "react";

/**
 * Whether the user is moving to the next or the previous page.
 */
export type PaginationDirection = "NEXT" | "PREVIOUS";

interface IPaginationProps {
  fetchPage: (
    startIndex: number,
    limit: number,
    direction: PaginationDirection
  ) => void;
  itemsOnCurrentPage: number;
  itemsPerPage: number;
  itemsTotal: number;
  startIndex: number;
}

/**
 * A component for offset-based pagination.
 *
 * The following properties need to be passed to the component.
 *
 * fetchPage:
 *     A method for fetching the next page. The method must expect a start index
 *     and a limit as its arguments.
 * itemsOnCurrentPage:
 *     The number of items on the current page. Unless the current page is the
 *     last page, this will be the same as the items per page.
 * itemsPerPage:
 *     The number of items to include on a page. For the last page the actual
 *     number of items may be less.
 * itemsTotal:
 *     The total number of items for all pages combined.
 * startIndex:
 *     The index of the first item on the current page. An index offset of 0,
 *     i.e. the start index for the first page is 0.
 */
const Pagination = (props: IPaginationProps) => {
  const {
    fetchPage,
    itemsOnCurrentPage,
    itemsPerPage,
    itemsTotal,
    startIndex
  } = props;
  const previousIndex = startIndex - itemsPerPage;
  const nextIndex = startIndex + itemsOnCurrentPage;

  return (
    <div>
      <button
        disabled={startIndex <= 0}
        className="pagination-previous"
        onClick={() => fetchPage(previousIndex, itemsPerPage, "PREVIOUS")}
      >
        Previous page
      </button>
      <a>
        <span className="pagination-ellipsis">{startIndex + 1}</span>
        <span className="pagination-ellipsis">&hellip;</span>
        <span className="pagination-ellipsis">
          {startIndex + itemsOnCurrentPage}
        </span>
        <span className="pagination-ellipsis">of</span>
        <span className="pagination-ellipsis">{itemsTotal}</span>
      </a>
      <button
        disabled={startIndex + itemsPerPage >= itemsTotal}
        className="pagination-next"
        onClick={() => fetchPage(nextIndex, itemsPerPage, "NEXT")}
      >
        Next page
      </button>
    </div>
  );
};

export default Pagination;
