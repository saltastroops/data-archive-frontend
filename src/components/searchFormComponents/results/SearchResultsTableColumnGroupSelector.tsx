import * as React from "react";
import styled from "styled-components";
import { LargeCheckbox } from "../../basicComponents/LargeCheckbox";
import SearchResultsTableColumn from "./ISearchResultsTableColumn";

const ColumnGroupHeading = styled.h2`
  font-size: 1em;
  font-weight: 700;
`;

interface ISearchResultsTableColumnGroupSelectorProps {
  category: string;
  columns: SearchResultsTableColumn[];
  onChange: (dataKey: string, visible: boolean) => void;
}

class SearchResultsTableColumnGroupSelector extends React.Component<
  ISearchResultsTableColumnGroupSelectorProps
> {
  render() {
    return (
      <div className="column-group-selector">
        <ColumnGroupHeading>{this.props.category}</ColumnGroupHeading>
        <ul>
          {this.props.columns.map(column => (
            <li key={column.dataKey}>
              <label>
                <LargeCheckbox
                  type="checkbox"
                  name={column.dataKey}
                  onChange={this.handleChange}
                  checked={column.visible}
                />
                {column.name}
              </label>
            </li>
          ))}
        </ul>
      </div>
    );
  }

  handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    this.props.onChange(e.target.name, e.target.checked);
  };
}

export default SearchResultsTableColumnGroupSelector;
