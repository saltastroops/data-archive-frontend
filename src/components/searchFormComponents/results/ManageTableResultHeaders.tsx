import React from "react";
import { LargeCheckbox } from "../../basicComponents/LargeCheckbox";
import ISearchResultsTableColumn from "./ISearchResultsTableColumn";

interface IManageTableResultHeadersProps {
  closeModal: () => void;
  displayed: ISearchResultsTableColumn[];
  notDisplayed: ISearchResultsTableColumn[];
  onChange: (
    a: ISearchResultsTableColumn[],
    b: ISearchResultsTableColumn[]
  ) => void;
}

interface IManageTableResultHeadersState {
  checked: ISearchResultsTableColumn[];
  left: ISearchResultsTableColumn[];
  leftChecked: ISearchResultsTableColumn[];
  right: ISearchResultsTableColumn[];
  rightChecked: ISearchResultsTableColumn[];
}

export default class ManageTableResultHeaders extends React.Component<
  IManageTableResultHeadersProps,
  IManageTableResultHeadersState
> {
  state: any = {
    checked: [],
    left: this.props.displayed,
    leftChecked: [],
    right: this.props.notDisplayed,
    rightChecked: [],
  };

  handleToggle = (value: any) => () => {
    const currentIndex = this.state.checked.indexOf(value);
    const newChecked = [...this.state.checked];

    if (currentIndex === -1) {
      newChecked.push(value);
    } else {
      newChecked.splice(currentIndex, 1);
    }

    this.setState({
      ...this.state,
      checked: newChecked,
      leftChecked: [...intersection(newChecked, this.state.left)],
      rightChecked: [...intersection(newChecked, this.state.right)],
    });
  };

  handleAllRight = () => {
    this.setState({
      ...this.state,
      left: [],
      leftChecked: [],
      right: this.state.right.concat(this.state.left),
      rightChecked: [],
    });
  };

  handleCheckedRight = () => {
    this.setState({
      ...this.state,
      checked: not(this.state.checked, this.state.leftChecked),
      left: not(this.state.left, this.state.leftChecked),
      leftChecked: [],
      right: this.state.right.concat(this.state.leftChecked),
      rightChecked: [],
    });
  };

  handleAllLeft = () => {
    this.setState({
      ...this.state,
      left: this.state.left.concat(this.state.right),
      leftChecked: [],
      right: [],
      rightChecked: [],
    });
  };

  handleCheckedLeft = () => {
    this.setState({
      ...this.state,
      checked: not(this.state.checked, this.state.rightChecked),
      left: this.state.left.concat(this.state.rightChecked),
      leftChecked: [],
      right: not(this.state.right, this.state.rightChecked),
      rightChecked: [],
    });
  };

  updateHeadersToDisplay = () => {
    this.props.onChange(this.state.left, this.state.right);
  };

  render() {
    const { closeModal } = this.props;
    const { checked, left, leftChecked, right, rightChecked } = this.state;

    const customList = (items: any) => (
      <div style={{ marginTop: "5%" }}>
        <ul>
          {items.map((value: any) => {
            const labelId = value.name;
            return (
              <li key={value.dataKey}>
                <LargeCheckbox
                  checked={checked.indexOf(value) !== -1}
                  tabIndex={-1}
                  onChange={this.handleToggle(value)}
                />{" "}
                {labelId}
              </li>
            );
          })}
        </ul>
      </div>
    );

    return (
      <div className="modal is-active">
        <div className="modal-background" onClick={closeModal} />
        <div className="modal-card column is-mobile is-half is-offset-one-quarter">
          <header className="modal-card-head">
            <p className="modal-card-title">Custom columns to show</p>
            <button
              className="delete"
              aria-label="close"
              onClick={closeModal}
            />
          </header>
          <section className="modal-card-body">
            <div className="columns is-mobile">
              <div className="column">
                <strong>Displayed</strong>
                {customList(left)}
              </div>
              <div className="column is-offset-1">
                <div style={{ marginTop: "25%" }} />
                <button
                  className={"button is-small"}
                  onClick={this.handleAllRight}
                  disabled={left.length === 0}
                  aria-label="move all right"
                >
                  ≫
                </button>
                <br />
                <button
                  className={"button is-small"}
                  onClick={this.handleCheckedRight}
                  disabled={leftChecked.length === 0}
                  aria-label="move selected right"
                >
                  &gt;
                </button>
                <br />
                <button
                  className={"button is-small"}
                  onClick={this.handleCheckedLeft}
                  disabled={rightChecked.length === 0}
                  aria-label="move selected left"
                >
                  &lt;
                </button>
                <br />
                <button
                  className={"button is-small"}
                  onClick={this.handleAllLeft}
                  disabled={right.length === 0}
                  aria-label="move all left"
                >
                  ≪
                </button>
              </div>
              <div className="column">
                <strong>Add to display</strong>
                {customList(right)}
              </div>
            </div>
          </section>
          <footer className="modal-card-foot">
            <button
              className="button is-primary"
              onClick={this.updateHeadersToDisplay}
            >
              Done
            </button>
          </footer>
        </div>
      </div>
    );
  }
}

// Needs relook
const not = (a: any, b: any) => {
  return a.filter((value: any) => b.indexOf(value) === -1);
};

// Needs relook
const intersection = (a: any, b: any) => {
  return a.filter((value: any) => b.indexOf(value) !== -1);
};
