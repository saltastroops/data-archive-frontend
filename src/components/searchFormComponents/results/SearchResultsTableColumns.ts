import moment from "moment";
import DataKeys from "./DataKeys";
import ISearchResultsTableColumn from "./ISearchResultsTableColumn";

/**
 * Get the search results table columns for an array of database column names.
 *
 * The returned array includes a dummy column (required for displaying purposes)
 * and any other columns which should always be included in the search results
 * table.
 *
 * Parameters:
 * -----------
 * dbColumns: string[]
 *     Database columns.
 *
 * Returns:
 * --------
 * The array of columns for the search results table.
 */
export function searchResultsTableColumns(
  dbColumns: string[]
): ISearchResultsTableColumn[] {
  const ignoredDataKeys = [
    DataKeys.DATA_FILE_ID,
    DataKeys.TARGET_TYPE_NUMERIC_CODE,
    DataKeys.TELESCOPE_OBSERVATION_ID,
    DataKeys.OBSERVATION_ID
  ];

  // The SearchResultsTable class sets the first column's width to the cart
  // column's width and puts the former underneath the latter. So
  // effectively the first column is ignored, and a dummy column needs to be
  // added to compensate for that.
  const columns = [
    { dataKey: "dummyName", name: "dummy", visible: true },
    {
      dataKey: DataKeys.OBSERVATION_NAME,
      name: "Observation",
      visible: true
    },
    ...sort(dbColumns.map(tableColumn))
  ];

  return columns.filter(column => !ignoredDataKeys.includes(column.dataKey));
}

/**
 * Return the search results table column for a data key.
 *
 * Parameters:
 * -----------
 * dataKey: string
 *     Data key.
 *
 * Returns:
 * --------
 * Search results table column.
 */
function tableColumn(dataKey: string): ISearchResultsTableColumn {
  switch (dataKey) {
    case DataKeys.DATA_CATEGORY:
      return { dataKey, name: "Data Category", visible: true };
    case DataKeys.DATA_FILE_FILENAME:
      return { dataKey, name: "File", visible: true };
    case DataKeys.DETECTOR_MODE:
      return { dataKey, name: "Detector Mode", visible: true };
    case DataKeys.FILTER:
      return { dataKey, name: "Filter", visible: true };
    case DataKeys.HRS_MODE:
      return { dataKey, name: "HRS Mode", visible: true };
    case DataKeys.INSTRUMENT_MODE:
      return { dataKey, name: "RSS Mode", visible: true };
    case DataKeys.INSTRUMENT_NAME:
      return { dataKey, name: "Instrument", visible: true };
    case DataKeys.OBSERVATION_NAME:
      return { dataKey, name: "Observation", visible: true };
    case DataKeys.OBSERVATION_NIGHT:
      return { dataKey, name: "Night", visible: true, format: formatTimestamp };
    case DataKeys.OBSERVATION_PUBLIC_FROM:
      return {
        dataKey,
        format: formatTimestamp,
        name: "Release Date",
        visible: true
      };
    case DataKeys.POLARIZATION_MODE:
      return { dataKey, name: "Polarimetry Pattern", visible: true };
    case DataKeys.PROPOSAL_CODE:
      return { dataKey, name: "Proposal", visible: true };
    case DataKeys.PROPOSAL_PI:
      return { dataKey, name: "PI", visible: true };
    case DataKeys.PROPOSAL_TITLE:
      return { dataKey, name: "Proposal Title", visible: true };
    case DataKeys.RSS_FABRY_PEROT_MODE:
      return { dataKey, name: "RSS FP Mode", visible: true };
    case DataKeys.RSS_GRATING:
      return { dataKey, name: "RSS Grating", visible: true };
    case DataKeys.TARGET_DECLINATION:
      return {
        dataKey,
        format: formatNumber(4),
        name: "Declination",
        visible: true
      };
    case DataKeys.TARGET_NAME:
      return { dataKey, name: "Target", visible: true };
    case DataKeys.TARGET_RIGHT_ASCENSION:
      return {
        dataKey,
        format: formatNumber(4),
        name: "Right Ascension",
        visible: true
      };
    case DataKeys.TARGET_TYPE_EXPLANATION:
      return { dataKey, name: "TargetType", visible: true };
    case DataKeys.TELESCOPE_NAME:
      return { dataKey, name: "Telescope", visible: true, width: 100 };
    default:
      return { dataKey, name: dataKey, visible: true };
  }
}

/**
 * Return a sorted copy of an array of columns.
 *
 * The original array is unaffected.
 *
 * Parameters:
 * -----------
 * columns: ISearchResultsTableColumn[]
 *     Columns.
 *
 * Returns:
 * --------
 * A sorted copy of the given array of columns.
 */
function sort(columns: ISearchResultsTableColumn[]) {
  const orderedDataKeys = [
    DataKeys.OBSERVATION_NAME,
    DataKeys.DATA_FILE_FILENAME,
    DataKeys.DATA_CATEGORY,
    DataKeys.OBSERVATION_NIGHT,
    DataKeys.TARGET_NAME,
    DataKeys.TARGET_RIGHT_ASCENSION,
    DataKeys.TARGET_DECLINATION,
    DataKeys.TARGET_TYPE_EXPLANATION,
    DataKeys.PROPOSAL_CODE,
    DataKeys.PROPOSAL_PI,
    DataKeys.PROPOSAL_TITLE,
    DataKeys.TELESCOPE_NAME,
    DataKeys.INSTRUMENT_NAME
  ];

  // Compare two columns. The following rules are used:
  //
  // * If the data keys of both columns are in orderedDataKeys, their order in
  //   this array is used.
  // * If only one column's data key is in orderedDataKeys, that column comes
  //   first.
  // * If none of the data keys is in orderedDataKeys, the column names are used
  //   for sorting
  const cmp = (
    column1: ISearchResultsTableColumn,
    column2: ISearchResultsTableColumn
  ) => {
    const index1 = orderedDataKeys.findIndex(key => column1.dataKey === key);
    const index2 = orderedDataKeys.findIndex(key => column2.dataKey === key);
    if (index1 !== -1 && index2 !== -1) {
      return index1 - index2;
    } else if (index1 !== -1 && index2 === -1) {
      return -1;
    } else if (index1 === -1 && index2 !== -1) {
      return 1;
    } else {
      return column1.name.localeCompare(column2.name, "en");
    }
  };

  return [...columns].sort(cmp);
}

/**
 * Returns a function which formats strings as a a number with a fixed number of
 * digits after the decimal point.
 *
 * Parameters:
 * -----------
 * n: number
 *     Number of digits after the decimal point.
 *
 * Returns:
 * --------
 * The format function.
 */
function formatNumber(n: number) {
  return (value: string) => Number(value).toFixed(n);
}

/**
 * Format a Unix timestamp as a string with the day, the month's abbreviated
 * name and the four-digit year, such as 1 Jan 2019 or 23 Aug 2020.
 *
 * Parameters:
 * -----------
 * t: string
 *     Unix timestamp.
 *
 * Returns:
 * --------
 * The formatted date string.
 */
function formatTimestamp(t: string) {
  return moment(parseInt(t, 10)).format("D MMM YYYY");
}
