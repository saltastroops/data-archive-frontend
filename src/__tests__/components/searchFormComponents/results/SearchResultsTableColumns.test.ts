import DataKeys from "../../../../components/searchFormComponents/results/DataKeys";
import { searchResultsTableColumns } from "../../../../components/searchFormComponents/results/SearchResultsTableColumns";

describe("SearchResultsTableColumns", () => {
  describe("searchResultsTableColumns", () => {
    it("should generate the correct table columns", () => {
      let dbColumns: string[] = [];
      expect(searchResultsTableColumns(dbColumns)).toMatchSnapshot();

      dbColumns = [DataKeys.INSTRUMENT_NAME];
      expect(searchResultsTableColumns(dbColumns)).toMatchSnapshot();

      dbColumns = [
        DataKeys.PROPOSAL_PI_FAMILY_NAME,
        DataKeys.TARGET_TYPE_NUMERIC_CODE,
        DataKeys.TARGET_DECLINATION,
        DataKeys.TELESCOPE_NAME,
        DataKeys.INSTRUMENT_NAME,
        DataKeys.PROPOSAL_CODE,
        DataKeys.TELESCOPE_OBSERVATION_ID,
        DataKeys.TARGET_RIGHT_ASCENSION,
        DataKeys.OBSERVATION_ID
      ];
      expect(searchResultsTableColumns(dbColumns)).toMatchSnapshot();
    });
  });
});
