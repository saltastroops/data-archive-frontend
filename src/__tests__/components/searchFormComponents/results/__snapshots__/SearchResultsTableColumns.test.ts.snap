// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsTableColumns searchResultsTableColumns should generate the correct table columns 1`] = `
Array [
  Object {
    "dataKey": "dummyName",
    "name": "dummy",
    "visible": true,
  },
  Object {
    "dataKey": "observationName",
    "name": "Observation",
    "visible": true,
  },
]
`;

exports[`SearchResultsTableColumns searchResultsTableColumns should generate the correct table columns 2`] = `
Array [
  Object {
    "dataKey": "dummyName",
    "name": "dummy",
    "visible": true,
  },
  Object {
    "dataKey": "observationName",
    "name": "Observation",
    "visible": true,
  },
  Object {
    "dataKey": "Instrument.instrumentName",
    "name": "Instrument",
    "visible": true,
  },
]
`;

exports[`SearchResultsTableColumns searchResultsTableColumns should generate the correct table columns 3`] = `
Array [
  Object {
    "dataKey": "dummyName",
    "name": "dummy",
    "visible": true,
  },
  Object {
    "dataKey": "observationName",
    "name": "Observation",
    "visible": true,
  },
  Object {
    "dataKey": "Target.rightAscension",
    "name": "Right Ascension",
    "visible": true,
  },
  Object {
    "dataKey": "Target.declination",
    "name": "Declination",
    "visible": true,
  },
  Object {
    "dataKey": "Proposal.proposalCode",
    "name": "Proposal",
    "visible": true,
  },
  Object {
    "dataKey": "Proposal.principalInvestigatorFamilyName",
    "name": "PI",
    "visible": true,
  },
  Object {
    "dataKey": "Instrument.instrumentName",
    "name": "Instrument",
    "visible": true,
  },
]
`;
